staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go

```diff
@@ -383,41 +384,123 @@ func (c *Controller) syncPool(ctx context.Context, poolName string) error {
 		if pool.Generation > generation {
 			generation = pool.Generation
 		}
-
-		// Right now all devices get published in a single slice.
-		// We simply pick the first one, if there is one, and copy
-		// it in preparation for updating it.
-		//
-		// TODO: support splitting across slices, with unit tests.
-		if len(slices) > 0 {
-			obsoleteSlices = append(obsoleteSlices, slices[1:]...)
-			slices = []*resourceapi.ResourceSlice{slices[0].DeepCopy()}
-		} else {
+		// If there are no slices, create an empty one.
+		if len(slices) == 0 {
 			slices = []*resourceapi.ResourceSlice{
 				{
 					ObjectMeta: metav1.ObjectMeta{
 						GenerateName: c.owner.Name + "-" + c.driver + "-",
 					},
 				},
 			}
+			slice := slices[0]
+			slice.OwnerReferences = []metav1.OwnerReference{{
+				APIVersion: c.owner.APIVersion,
+				Kind:       c.owner.Kind,
+				Name:       c.owner.Name,
+				UID:        c.owner.UID,
+				Controller: ptr.To(true),
+			}}
+			slice.Spec.Driver = c.driver
+			slice.Spec.Pool.Name = poolName
+			slice.Spec.Pool.Generation = generation
+			slice.Spec.Pool.ResourceSliceCount = 1
+			slice.Spec.NodeName = nodeName
+			slice.Spec.NodeSelector = pool.NodeSelector
+			slice.Spec.AllNodes = pool.NodeSelector == nil && nodeName == ""
+			slice.Spec.Devices = pool.Devices
 		}
-
-		slice := slices[0]
-		slice.OwnerReferences = []metav1.OwnerReference{{
-			APIVersion: c.owner.APIVersion,
-			Kind:       c.owner.Kind,
-			Name:       c.owner.Name,
-			UID:        c.owner.UID,
-			Controller: ptr.To(true),
-		}}
-		slice.Spec.Driver = c.driver
-		slice.Spec.Pool.Name = poolName
-		slice.Spec.Pool.Generation = generation
-		slice.Spec.Pool.ResourceSliceCount = 1
-		slice.Spec.NodeName = nodeName
-		slice.Spec.NodeSelector = pool.NodeSelector
-		slice.Spec.AllNodes = pool.NodeSelector == nil && nodeName == ""
-		slice.Spec.Devices = pool.Devices
+		needToAdd, needToDelete := getDevicesToAddAndDelete(pool, slices)
+		// there are changes in a pool, re-publish all slices of that pool with an increased pool.generation
+		if len(needToAdd) > 0 || len(needToDelete) > 0 {
```

Comment:
When the pool device changes, do we increase the number of slice generations? Because I don't see the logic of increasing the number of pool generations. @pohly

Reply:
If the only change is modifying exactly one slice, then we don't need to bump the generation. The recipients either has seen that new slice or they haven't.

But if the changes affect more than one slice, then bumping the generation and re-publishing the entire pool is safer.